### ISSUE TITLE
Rule idea: `no-invalid-descriptor-declaration`

### ISSUE THREAD

**Issue by silverwind** [stylelint/stylelint#6859]
### What is the problem you're trying to solve?

Using the `var()` function inside a `@` rule block is a CSS syntax error that should be detectable:

```css
@font-face {
  font-family: foo;
  font-weight: var(--font-weight-normal); /* Syntax Error */
}
```

Firefox will raise a `Unknown descriptor ‘var(’ in @font-face rule.  Skipped to next declaration.`.

### What solution would you like to see?

A rule that detects such cases. I dug through the CSS specs and found:

> [Block at-rules](https://drafts.csswg.org/css-syntax-3/#block-at-rule) will typically contain a collection of (generic or [at-rule](https://drafts.csswg.org/css-syntax-3/#at-rule)–specific) at-rules, [qualified rules](https://drafts.csswg.org/css-syntax-3/#qualified-rule), and/or [descriptor declarations](https://drafts.csswg.org/css-syntax-3/#css-descriptor-declarations) subject to limitations defined by the at-rule. Descriptors are similar to [properties](https://drafts.csswg.org/css-cascade-5/#css-property) (and are declared with the same syntax) but are associated with a particular type of at-rule rather than with elements and boxes in the tree.

> The [var()](https://drafts.csswg.org/css-variables/#funcdef-var) function can be used in place of any part of a value in any property on an element. The var() function can not be used as property names, selectors, or anything else besides property values. (Doing so usually produces invalid syntax, or else a value whose meaning has no connection to the variable.)

I'm sure there are more functions besides `var` that would be invalid for a descriptor declaration, but a rule that just warns for `var` would be a good start.

**Comment by silverwind** [comment]
If stylelint had a proper parsers like proposed in https://github.com/stylelint/stylelint/issues/6853, this case wouldn't even need a rule, as it would trip that parser.

**Comment by ybiquitous** [comment]
@silverwind Thanks for the proposal. Could you please comment on a link to the CSS spec you quoted?

Also, does this syntax error occur with other at-rules?

**Comment by silverwind** [comment]
Spec links are:

https://drafts.csswg.org/css-syntax-3/
https://drafts.csswg.org/css-variables/

As for `@` rules, it seems it does not affect all of them. `@page` accepts `var`, but `@font-face` and `@counter-style` does not. So it seems the spec is not entirely true either. Happy to close this off as "too hard" if you like.

**Comment by alexander-akait** [comment]
@silverwind yeah, I think it is more about grammar parsing than linting, so I hope in future we resolve problems with our parser and will catch this on the parser side (or will have rule to apply grammar parsing to each things in your CSS)

**Comment by silverwind** [comment]
Agree, syntax errors are topics for a parser, not a linter.

**Comment by romainmenke** [comment]
@ybiquitous We might want to keep this one open and add a rule for it regardless of anything that might be done for parsing CSS in general?
This seems like a common mistake that can easily be detected.

Thoughts?

**Comment by ybiquitous** [comment]
@romainmenke Surely, this rule might help people to notice mistakes of `var()` etc. in at-rules.

But, I'm concerning about the @silverwind's [comment](https://github.com/stylelint/stylelint/issues/6859#issuecomment-1563330538):

> So it seems the spec is not entirely true either. Happy to close this off as "too hard" if you like.

Can we solve this problem?

---

Anyway, I agree with reopening this and keeping it until we find a good solution. 👍🏼

**Comment by romainmenke** [comment]
If I recall correctly this is the difference between descriptors and properties.
`@font-face`, `@property`, `@counter-style` all take a list of descriptors, whereas `@page` takes a list of declarations (property / value pairs).

https://www.w3.org/TR/css-variables-1/#using-variables

> The value of a custom property can be substituted into the value of another **property** with the var() function. The syntax of var() is:

https://www.w3.org/TR/css-fonts-4/#font-face-rule

**Comment by ybiquitous** [comment]
Ah, I see! Some at-rules accept properties, while some at-rules accept descriptors.

If so, we can likely manage at-rules accepting descriptors in [`reference/atKeywords.js`](https://github.com/stylelint/stylelint/blob/40b4a6daa6168201e55eedae2acef431e216978b/lib/reference/atKeywords.js).

**Comment by ybiquitous** [comment]
If this rule is limited to at-rule, starting the rule name with `at-rule-` might be better. 🤔 

Ref:
- https://stylelint.io/developer-guide/rules#define-the-rule
- https://stylelint.io/user-guide/rules/#invalid

**Comment by silverwind** [comment]
Agree, renamed title to `at-rule-no-invalid-descriptor`.

**Comment by ybiquitous** [comment]
Okay, I suggest a blueprint:

- Name: `at-rule-no-invalid-descriptor`
- Description: Disallow invalid descriptors within at-rules.
- Message: `Unexpected invalid descriptor within "@<at-rule>"`
- Primary option: `true`
- Secondary options: n/a
- Category: Avoid errors > Invalid
- Included in sharable configs?: No. But if it matured, we would include it in the standard config.

Example:

```css
@font-face {
  font-weight: var(--font-weight-normal);
/*             ^~~~~~~~~~~~~~~~~~~~~~~~~
               Unexpected invalid descriptor within "@<at-rule>"
 */
}
```

```json
{
  "at-rule-no-invalid-descriptor": true
}
```

Please feel free to give me any feedback.

**Comment by Mouvedia** [comment]
> no-invalid-descriptor

Not sure if this is the right name because
```css
:root {
--lol : 6in;
}

@page {
  size : 4in var(--lol); /* size is a descriptor */
}
```
seems to be valid.

**Comment by silverwind** [comment]
Right, it's only refering to the right side, so `at-rule-no-invalid-descriptor-value`.

**Comment by Mouvedia** [comment]
What I was trying to say is that the scope of `no-invalid-descriptor-value` might require to list every single cases in the documentation.
i.e. right now it's only `var()` that has been identified but there are probably other cases

**Comment by silverwind** [comment]
Ah, misunderstood you. If `var()` is valid in some constallations, it'll make the rule more complicated. I wonder if there is data around somewhere which CSS functions are considered invalid in descriptor values, there's certainly more than just `var()`.

Regarding rule name, I think `at-rule-no-invalid-descriptor-value` may be a better name because then `at-rule-no-invalid-descriptor` could be used for a future rule that validates the descriptor names  because at least some at-rules like `@font-face` have a pre-defined list of valid descriptor names.

**Comment by Mouvedia** [comment]
We had the inverse problem with `function-calc-no-invalid` which scope ended up being too small.
We are trying to limit the future maintenance by picking the name precisely.
Ill let @jeddy3 chime in before continuing with the bikeshedding.

**Comment by romainmenke** [comment]
Maybe it can be approached from a different perspective?

Part of the issue here is that `var()` is expected by many to just work everywhere.
Especially for people coming from preprocessors like scss.

A rule that warns on invalid locations of `var()` functions could cover :
- used in `@font-face` descriptor values
- used in `@media` feature values
- used in selectors
- ...

I think this is also easier to test and verify so that it is correct with the CSS specification.

**Comment by ybiquitous** [comment]
Hum, @romainmenke's suggestion sounds good. So, how about `function-var-no-invalid-location`?

**Comment by jeddy3** [comment]
> Part of the issue here is that var() is expected by many to just work everywhere.

That's a good summary.

> I think this is also easier to test and verify so that it is correct with the CSS specification.

I agree.

> So, how about `function-var-no-invalid-location`?

SGTM. It'll tie into the "The `var()` function can not be used as property names, selectors, or anything else besides property values" part of the spec that @silverwind [mentioned above](https://github.com/stylelint/stylelint/issues/6859#issue-1722817603).

I had thought `env` and `var` were similar and so a rule would apply to them both, but it turns out [`env` can be used in more places](https://developer.mozilla.org/en-US/docs/Web/CSS/env):

> In addition, unlike custom properties, which cannot be used outside of declarations, the env() function can be used in place of any part of a property value, or any part of a descriptor (e.g. in [Media query rules](https://developer.mozilla.org/en-US/docs/Web/CSS/@media)). As the spec evolves, it may also be usable in other places such as selectors.

Our closest existing rule is [no-invalid-position-at-import-rule](https://stylelint.io/user-guide/rules/no-invalid-position-at-import-rule), so called because it's scoped to the whole source like this rule.

Suggested blueprint:

- Name: `no-invalid-position-var-function`
- Description: Disallow invalid position `var()` functions.
- Message: Unexpected invalid position var() function
- Primary option: `true`
- Secondary options: n/a
- Category: Avoid errors > Invalid
- Included in sharable configs?: Yes, recommended.
- Expanded description:

The [`var()`](https://drafts.csswg.org/css-variables/#funcdef-var) function can be used in place of any part of a value in any property on an element. The `var()` function can not be used as property names, selectors, or anything else besides property values.

This rule checks:

- at-rule descriptor names and values
- media queries
- property names
- selectors

---

We'll want to ensure we don't [needlessly call any of the constructor parsers](https://github.com/stylelint/stylelint/issues/6869#issuecomment-1605428748) when checking any of the above.

According to the [CSS tree reference](https://csstree.github.io/docs/syntax/#report&q=ZGljdC5bdHlwZT0iQXRydWxlIl0%3D), the following at-rule have descriptors:

- `counter-style`
- `font-face`
- `page`
- `property`
- `viewport`

Shall I label as ready to implement?

**Comment by ybiquitous** [comment]
I agree with @jeddy3's suggestion 👍🏼
